From patchwork Wed Feb 11 03:32:48 2026
Content-Type: text/plain; charset="utf-8"
MIME-Version: 1.0
Content-Transfer-Encoding: 7bit
X-Patchwork-Submitter: James Hilliard <james.hilliard1@gmail.com>
X-Patchwork-Id: 14419299
Return-Path: 
 <linux-arm-kernel-bounces+linux-arm-kernel=archiver.kernel.org@lists.infradead.org>
X-Spam-Checker-Version: SpamAssassin 3.4.0 (2014-02-07) on
	aws-us-west-2-korg-lkml-1.web.codeaurora.org
Received: from bombadil.infradead.org (bombadil.infradead.org
 [198.137.202.133])
	(using TLSv1.2 with cipher ECDHE-RSA-AES256-GCM-SHA384 (256/256 bits))
	(No client certificate requested)
	by smtp.lore.kernel.org (Postfix) with ESMTPS id 66F06EB595E
	for <linux-arm-kernel@archiver.kernel.org>;
 Wed, 11 Feb 2026 03:33:19 +0000 (UTC)
DKIM-Signature: v=1; a=rsa-sha256; q=dns/txt; c=relaxed/relaxed;
	d=lists.infradead.org; s=bombadil.20210309; h=Sender:List-Subscribe:List-Help
	:List-Post:List-Archive:List-Unsubscribe:List-Id:Content-Transfer-Encoding:
	MIME-Version:Message-ID:Date:Subject:Cc:To:From:Reply-To:Content-Type:
	Content-ID:Content-Description:Resent-Date:Resent-From:Resent-Sender:
	Resent-To:Resent-Cc:Resent-Message-ID:In-Reply-To:References:List-Owner;
	bh=lqL9RDMHFR3WoHQnRB4Y8oAV8tGCe4j0xxAYm/MQPvM=; b=nyJRJRihLpqqiwUeTBqkTi9a+5
	bHUmzfpG0iT/xWDC0OnGmNzUAPpOaSU/ExNLp/XYo2ZQSjBFDUvZUOfBihjutJTPjZps9BpQyet9a
	zlrDTO4xNmprO2QDnMvTlKmraCZglGRMB5q30xRCnzHEGUQYEUEDPEXJFAu48nJq/OlHYcMq+v0mY
	0JELEbw7jl8/iQvnP2LtiDhd7qYuBFUPrBX5HiMBkCJyes5bcbqDbCCex3740ETxbK6r8QsPZkuEu
	/1uC8MC4pXr5RhUjfkNTsoMxN95rJ86t8nStBJunKVKgzKoyV9wzQFwTGVISTYOIbj+JzkOikdTSu
	fEpP6yWg==;
Received: from localhost ([::1] helo=bombadil.infradead.org)
	by bombadil.infradead.org with esmtp (Exim 4.98.2 #2 (Red Hat Linux))
	id 1vq0yT-000000009v5-1XSj;
	Wed, 11 Feb 2026 03:33:09 +0000
Received: from mail-ot1-x333.google.com ([2607:f8b0:4864:20::333])
	by bombadil.infradead.org with esmtps (Exim 4.98.2 #2 (Red Hat Linux))
	id 1vq0yR-000000009uO-0LTl
	for linux-arm-kernel@lists.infradead.org;
	Wed, 11 Feb 2026 03:33:08 +0000
Received: by mail-ot1-x333.google.com with SMTP id
 46e09a7af769-7d1890f7ee4so3961143a34.0
        for <linux-arm-kernel@lists.infradead.org>;
 Tue, 10 Feb 2026 19:33:06 -0800 (PST)
DKIM-Signature: v=1; a=rsa-sha256; c=relaxed/relaxed;
        d=gmail.com; s=20230601; t=1770780785; x=1771385585;
 darn=lists.infradead.org;
        h=content-transfer-encoding:mime-version:message-id:date:subject:cc
         :to:from:from:to:cc:subject:date:message-id:reply-to;
        bh=lqL9RDMHFR3WoHQnRB4Y8oAV8tGCe4j0xxAYm/MQPvM=;
        b=nR5oRWsfV53TaK6WqVz+Hfrs4kPSHGtft7LSMrGYiiWKIJhpHRNMsJ1/fAh9paFtEO
         nrrvnJ5Z7kDKq2DKDCjXdaqa2QGrKdEJYCiqd5XGBaOv0+a0OHNaDx8gFNwczZ0vNTfk
         pzsYhTob2b1t13Cv2reiuzG8FbehEbOgDLCnYsULYNaDsFQ5PFElPKz0xGdMR2UnWC44
         eANFNUEUA6TpBtRQE7qR5MhWDfkEKUbCxmacQRcTXdKBPp01SkxOAnSSC463LbmfNWj0
         Rqgbd+x/Akf62E2kODhRrhGek3jQnmrrrhLkgChB4+Y/cX15MXBX7NgvpJ/RYPpYPzQD
         timQ==
X-Google-DKIM-Signature: v=1; a=rsa-sha256; c=relaxed/relaxed;
        d=1e100.net; s=20230601; t=1770780785; x=1771385585;
        h=content-transfer-encoding:mime-version:message-id:date:subject:cc
         :to:from:x-gm-gg:x-gm-message-state:from:to:cc:subject:date
         :message-id:reply-to;
        bh=lqL9RDMHFR3WoHQnRB4Y8oAV8tGCe4j0xxAYm/MQPvM=;
        b=kbGBwvgGfkGkxyTvqzgdZ9m5Zb1SJlFtuY/jui11ma0GXA8KKK4NCQUGh1hpWkNk+R
         1BMnFFRsHR2NiYCiWx7u32L1rdyWYlu66zQ3PbY1UvD+7ahjJ2/Fe6NnOBJvdgz2aPQV
         PZb7n4+mLhI8dU0xtCCYUTTu5iIDQAoHhS6amCgx54P7T89dBVhXXPxw3WEwJqkUsMIg
         SRkV93PlDsM1UbbqGXP3VmaJBjL+Na5q/kVEiOBpu9aki5ergDEeMiLSm2LStunTkUbv
         ckcHJmsKP7kpIxaowry8Ww2IzMUNWWyBq98l2ItseXYP1XQ556y4pxt1pdOFmX8HVHnA
         foGQ==
X-Forwarded-Encrypted: i=1;
 AJvYcCUQJM/kbv3KQnkzmb8RjNm4xQmZyxlI2PDA8v71DowkNMD9vGqC2VkikveXd3MGpA6oopla4Gjj3pVqGApBDmvJ@lists.infradead.org
X-Gm-Message-State: AOJu0YwAHfe9UGU8UHlU7mFwNNdDGg403VUnxBoORwwha9zwKtKmUvoC
	6SWOSdx0DsVJYgA03RBwoAeRAM0viNg25vstnaUCFqLdevLK2PYX+yZq
X-Gm-Gg: AZuq6aKLBUX8b6JnHfi2FMPSj7FfxY01BD0vr7/A//lEX05lBduoNHo6VTPOaVwa1xa
	0SrGinP9xGV02x5u4jFmCTsKSMMhYReO7xsppfKCQqbwcoAB6ildEen2jfRtPToFpQGz+gtLbY0
	1NvHJ/5RozXZlK15HwKXSP7cGR14zSwxSLra93tVJcdNSGLKcurZTjJikWEBnOUtoD0dfwO9q3N
	KxM14xUKMhYglemK2Fnwtmzy5Dmvk0uQ9X9YHMmfFx0TjewAK4CedZlC8tYb382fySN6HE4TZir
	HBWeO0meXNbxWd3ux8vptd9QL3PubrTaPEsSalVUVoMHPPXXhDxwySuXuOld5bPAiNiJB9X5BDM
	4qhNU5Y8FbgxvLCGyGSefcJqSoSK31jksO63ajNeorY8KQ/rUp2Y7tDpFVeqL3fHLq+B6aXl3hG
	HXo/NaBnxkSBdVdAvaptq6Vwz/90U9B+tdwl/4C0s85v4Pfbgl/5WLDklcgTCQvdvM4j3f9v2xJ
	J9/nO4km3sRjoHJ/MUWv1CPXz7EZtmUwy0m9INfCbenGu2M9nDSC4Vjt/iF5eE=
X-Received: by 2002:a05:6830:4494:b0:7d1:586a:32a2 with SMTP id
 46e09a7af769-7d4a73cc7c1mr381970a34.0.1770780785557;
        Tue, 10 Feb 2026 19:33:05 -0800 (PST)
Received: from james-x399.localdomain (71-218-105-26.hlrn.qwest.net.
 [71.218.105.26])
        by smtp.gmail.com with ESMTPSA id
 46e09a7af769-7d4a75309a6sm438236a34.3.2026.02.10.19.33.04
        (version=TLS1_3 cipher=TLS_AES_256_GCM_SHA384 bits=256/256);
        Tue, 10 Feb 2026 19:33:05 -0800 (PST)
From: James Hilliard <james.hilliard1@gmail.com>
To: linux-sunxi@lists.linux.dev,
	linux-gpio@vger.kernel.org
Cc: James Hilliard <james.hilliard1@gmail.com>,
	Linus Walleij <linusw@kernel.org>,
	Chen-Yu Tsai <wens@kernel.org>,
	Jernej Skrabec <jernej.skrabec@gmail.com>,
	Samuel Holland <samuel@sholland.org>,
	Bartosz Golaszewski <brgl@kernel.org>,
	linux-arm-kernel@lists.infradead.org,
	linux-kernel@vger.kernel.org
Subject: [PATCH 1/1] pinctrl: sunxi: add GPIO get_direction callback
Date: Tue, 10 Feb 2026 20:32:48 -0700
Message-ID: <20260211033249.2770281-1-james.hilliard1@gmail.com>
X-Mailer: git-send-email 2.43.0
MIME-Version: 1.0
X-CRM114-Version: 20100106-BlameMichelson ( TRE 0.8.0 (BSD) ) MR-646709E3 
X-CRM114-CacheID: sfid-20260210_193307_183420_F372974A 
X-CRM114-Status: GOOD (  13.88  )
X-BeenThere: linux-arm-kernel@lists.infradead.org
X-Mailman-Version: 2.1.34
Precedence: list
List-Id: <linux-arm-kernel.lists.infradead.org>
List-Unsubscribe: 
 <http://lists.infradead.org/mailman/options/linux-arm-kernel>,
 <mailto:linux-arm-kernel-request@lists.infradead.org?subject=unsubscribe>
List-Archive: <http://lists.infradead.org/pipermail/linux-arm-kernel/>
List-Post: <mailto:linux-arm-kernel@lists.infradead.org>
List-Help: <mailto:linux-arm-kernel-request@lists.infradead.org?subject=help>
List-Subscribe: 
 <http://lists.infradead.org/mailman/listinfo/linux-arm-kernel>,
 <mailto:linux-arm-kernel-request@lists.infradead.org?subject=subscribe>
Sender: "linux-arm-kernel" <linux-arm-kernel-bounces@lists.infradead.org>
Errors-To: 
 linux-arm-kernel-bounces+linux-arm-kernel=archiver.kernel.org@lists.infradead.org

Implement sunxi_pinctrl_gpio_get_direction() and wire it into
the sunxi gpio_chip setup.

The new callback reads the pin mux register and compares the mux
value against the pin descriptor gpio_in and gpio_out functions
to report GPIO_LINE_DIRECTION_IN or GPIO_LINE_DIRECTION_OUT.
If the pin is muxed to irq, report it as input.

Signed-off-by: James Hilliard <james.hilliard1@gmail.com>
---
 drivers/pinctrl/sunxi/pinctrl-sunxi.c | 32 +++++++++++++++++++++++++++
 1 file changed, 32 insertions(+)

diff --git a/drivers/pinctrl/sunxi/pinctrl-sunxi.c b/drivers/pinctrl/sunxi/pinctrl-sunxi.c
index 0fb057a07dcc..424f23be27b2 100644
--- a/drivers/pinctrl/sunxi/pinctrl-sunxi.c
+++ b/drivers/pinctrl/sunxi/pinctrl-sunxi.c
@@ -995,6 +995,37 @@ static int sunxi_pinctrl_gpio_direction_output(struct gpio_chip *chip,
 					    chip->base + offset, false);
 }
 
+static int sunxi_pinctrl_gpio_get_direction(struct gpio_chip *chip,
+					    unsigned int offset)
+{
+	struct sunxi_pinctrl *pctl = gpiochip_get_data(chip);
+	struct sunxi_desc_function *in, *out, *irq;
+	u32 reg, shift, mask, val;
+	u16 pin = chip->base + offset;
+
+	in = sunxi_pinctrl_desc_find_function_by_pin(pctl, pin, "gpio_in");
+	out = sunxi_pinctrl_desc_find_function_by_pin(pctl, pin, "gpio_out");
+	if (!in || !out)
+		return -EINVAL;
+
+	irq = sunxi_pinctrl_desc_find_function_by_pin(pctl, pin, "irq");
+
+	sunxi_mux_reg(pctl, offset, &reg, &shift, &mask);
+	val = (readl(pctl->membase + reg) & mask) >> shift;
+
+	if (val == in->muxval)
+		return GPIO_LINE_DIRECTION_IN;
+
+	if (val == out->muxval)
+		return GPIO_LINE_DIRECTION_OUT;
+
+	/* IRQ function is effectively input. */
+	if (irq && val == irq->muxval)
+		return GPIO_LINE_DIRECTION_IN;
+
+	return -EINVAL;
+}
+
 static int sunxi_pinctrl_gpio_of_xlate(struct gpio_chip *gc,
 				const struct of_phandle_args *gpiospec,
 				u32 *flags)
@@ -1603,6 +1634,7 @@ int sunxi_pinctrl_init_with_flags(struct platform_device *pdev,
 	pctl->chip->set_config = gpiochip_generic_config;
 	pctl->chip->direction_input = sunxi_pinctrl_gpio_direction_input;
 	pctl->chip->direction_output = sunxi_pinctrl_gpio_direction_output;
+	pctl->chip->get_direction = sunxi_pinctrl_gpio_get_direction;
 	pctl->chip->get = sunxi_pinctrl_gpio_get;
 	pctl->chip->set = sunxi_pinctrl_gpio_set;
 	pctl->chip->of_xlate = sunxi_pinctrl_gpio_of_xlate;
